notebooks/Threshold Optimization Postprocessing for Binary Classification.ipynb

```diff
@@ -0,0 +1,458 @@
+    "        x_tick_labels_basic.append(group)\n",
+    "    \n",
+    "        if y is not None:\n",
+    "            positive_indices[group] = sensitive_features.index[(sensitive_features == group) & (y == 1)]\n",
+    "            negative_indices[group] = sensitive_features.index[(sensitive_features == group) & (y == 0)]\n",
+    "            prob_1 = sum(y_pred[positive_indices[group]])/len(positive_indices[group])\n",
+    "            prob_0 = sum(y_pred[negative_indices[group]])/len(negative_indices[group])\n",
+    "            print(\"P[recidivism predicted | {}, recidivism]    {}= {}\".format(group, \" \"*(max_group_length-len(group)) , prob_1))\n",
+    "            print(\"P[recidivism predicted | {}, no recidivism] {}= {}\".format(group, \" \"*(max_group_length-len(group)), prob_0))\n",
+    "\n",
+    "            plt.bar(n_groups + 1 + 2 * index, prob_1, color=color[index])\n",
+    "            plt.bar(n_groups + 2 + 2 * index, prob_0, color=color[index])\n",
+    "            x_tick_labels_by_label.extend([\"{} recidivism\".format(group), \"{} no recidivism\".format(group)])\n",
+    "    \n",
+    "    x_tick_labels = x_tick_labels_basic + x_tick_labels_by_label\n",
+    "    plt.xticks(range(1, len(x_tick_labels)+1), x_tick_labels, rotation=45, horizontalalignment=\"right\")\n",
+    "        \n",
+    "show_proportions(X_train, sensitive_features_train, y_train, description=\"original training data:\", plot_row_index=1)\n",
+    "show_proportions(X_train, sensitive_features_train, unconstrained_predictor.predict(X_train), y_train, description=\"fairness-unaware prediction on training data:\", plot_row_index=2)\n",
+    "show_proportions(X_test, sensitive_features_test, y_test, description=\"original test data:\", plot_row_index=3)\n",
+    "show_proportions(X_test, sensitive_features_test, unconstrained_predictor.predict(X_test), y_test, description=\"fairness-unaware prediction on test data:\", plot_row_index=4)\n",
+    "plt.show()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We notice a stark contrast in the predictions with males being a lot more likely to be predicted to get approved, similar to the original training data. However, there's even a disparity between the subgroup of males and females that are approved. When considering only the samples labeled with \"loan not repaid\" males are much more likely to be predicted to get approved than females. The test data shows a similar disparity."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from fairlearn.widget import FairlearnDashboard\n",
+    "\n",
+    "predicted_ys = [unconstrained_predictor.predict(X_test).tolist()]\n",
+    "sensitive_features_mapped = list(map(lambda x: [x], sensitive_features_test.values))\n",
+    "\n",
+    "FairlearnDashboard(sensitive_features=sensitive_features_mapped,\n",
```

Comment:
For some reason, when I say 'sex' as the sensitive feature in the dashboard, the two categories which come back appear to be the binary race ones.

Reply:
This changed this morning. The feature_names need to be ["race"].